x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/XContentObjectTransformer.java

```diff
@@ -0,0 +1,82 @@
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.DeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchModule;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * This is a utility class that allows simple one-to-one transformations between an ToXContentObject type
+ * to and from other supported objects.
+ *
+ * @param <T> The type of the object that we will be transforming to/from
+ */
+public class XContentObjectTransformer<T extends ToXContentObject> {
+
+    private final NamedXContentRegistry registry;
+    private final CheckedFunction<XContentParser, T, IOException> parserFunction;
+
+    // We need this registry for parsing out Aggregations and Searches
+    private static NamedXContentRegistry searchRegistry;
+    static {
+        SearchModule searchModule = new SearchModule(Settings.EMPTY, false, Collections.emptyList());
+        searchRegistry = new NamedXContentRegistry(searchModule.getNamedXContents());
+    }
+
+    public static XContentObjectTransformer<AggregatorFactories.Builder> aggregatorTransformer() {
+        return new XContentObjectTransformer<>(searchRegistry, (p) -> {
+            // Serializing a map creates an object, need to skip the start object for the aggregation parser
+            assert(XContentParser.Token.START_OBJECT.equals(p.nextToken()));
+            return AggregatorFactories.parseAggregators(p);
+        });
+    }
+
+    public static XContentObjectTransformer<QueryBuilder> queryBuilderTransformer() {
+        return new XContentObjectTransformer<>(searchRegistry, AbstractQueryBuilder::parseInnerQueryBuilder);
+    }
+
+    XContentObjectTransformer(NamedXContentRegistry registry, CheckedFunction<XContentParser, T, IOException> parserFunction) {
+        this.parserFunction = parserFunction;
+        this.registry = registry;
+    }
+
+    public T fromMap(Map<String, Object> stringObjectMap) throws IOException {
+        try(XContentBuilder xContentBuilder = XContentFactory.jsonBuilder().map(stringObjectMap);
+            XContentParser parser = XContentType.JSON
+                .xContent()
+                .createParser(registry,
+                    DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
```

Comment:
I think this should use a custom implementation of `DeprecationHandler` that:

1. Starts by calling the equivalent method on `LoggingDeprecationHandler.INSTANCE` so that deprecations get logged
2. Also builds up any deprecation messages in a `List` that can be got at the end of the parsing

If someone is using deprecated search syntax in a datafeed then we _don't_ want to refuse to use it in the version  where it's just deprecated.  In normal operation we want to log it to the deprecation log, and in the upgrade assistant we'll want to report the problems in some other way, hence the suggestion for a custom deprecation logger that can do both these things.

In the case where we're parsing syntax that was deprecated in the previous major version and is no longer understood, this method will of course throw an exception - we'll have to handle this in the calling code.